### DISCUSSION TITLE
BlueCat provider

### DISCUSSION THREAD

**Discussion by dittman1** [go-acme/lego#1625] [Q&A] [answered]
I've downloaded lego 4.6.  I'm using BlueCat 9.4.  I've set the environment variables but when I try to test I get an error "UNAUTHORIZED USER".  I've verified the username and password is correct and the account is enabled for API.  Here's what I'm getting along with the environment variables (with dummy values substituted in the variables and output):

```
BLUECAT_PASSWORD=mypassword
BLUECAT_DNS_VIEW=myview
BLUECAT_USER_NAME=myusername
BLUECAT_CONFIG_NAME=myconfig
BLUECAT_SERVER_URL=http://bam.mydomain/Services/REST/v1/
BLUECAT_TTL=30

# ./lego --server="https://acme-staging-v02.api.letsencrypt.org/directory" --dns="bluecat" --domains="host.mydomain" --email="my email" run
2022/04/12 00:14:03 [INFO] [mydomain] acme: Obtaining bundled SAN certificate
2022/04/12 00:14:03 [INFO] [mydomain] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/2162808988
2022/04/12 00:14:03 [INFO] [mydomain] acme: Could not find solver for: tls-alpn-01
2022/04/12 00:14:03 [INFO] [mydomain] acme: Could not find solver for: http-01
2022/04/12 00:14:03 [INFO] [mydomain] acme: use dns-01 solver
2022/04/12 00:14:03 [INFO] [mydomain] acme: Preparing to solve DNS-01
2022/04/12 00:14:03 [INFO] [mydomain] acme: Cleaning DNS-01 challenge
2022/04/12 00:14:03 [WARN] [mydomain] acme: cleaning up failed: bluecat: request failed with HTTP status code 401
 Full message: "UNAUTHORIZED USER" 
2022/04/12 00:14:03 [INFO] Deactivating auth: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/2162808988
2022/04/12 00:14:03 Could not obtain certificates:
	error: one or more domains had a problem:
[host.mydomain] [host.mydomain] acme: error presenting token: bluecat: request failed with HTTP status code 401
 Full message: "UNAUTHORIZED USER"
```

Also for the certificate how do I specify an additional FQDN for the cert (for example, www.mydomain)?

**Comment by dittman1** [comment]
I'm making progress, it turns out the URL should just be "http://bam.mydomain".  With that it I can run it and it adds the TXT record but I got an error that the time limit was exceeded for the record to propagate even though I was able to verify the TXT record was resolving.  More to dig into.

**Comment by dittman1** [comment]
I'm still getting time limit exceeded, sometimes with the error that the record can't be resolved, other times that it doesn't match the expected value of the TXT record.  Part of that is each run creates the TXT record but fails cleanup:

`2022/04/13 09:31:29 [WARN] [bam.mydomain] acme: cleaning up failed: bluecat: request failed with HTTP status code 404
 Full message: {"status":404,"reason":"Not Found","code":"ResourceNotFound","message":"The requested resource was not found"} 
` 
I always have to manually delete the TXT record.

**Comment by ldez** [comment, marked as answer]
Related to https://github.com/go-acme/lego/issues/1626 fixed by https://github.com/go-acme/lego/pull/1627
Available in v4.7.0
